host/workflowsidinternalratelimit_test.go

```diff
@@ -0,0 +1,231 @@
+		Identity:                            testIdentity,
+		WorkflowIDReusePolicy:               types.WorkflowIDReusePolicyTerminateIfRunning.Ptr(),
+	}
+
+	ctx, cancel := createContext()
+	defer cancel()
+
+	we, err := s.engine.StartWorkflowExecution(ctx, request)
+	s.NoError(err)
+
+	s.Logger.Info("StartWorkflowExecution", tag.WorkflowRunID(we.RunID))
+
+	workflowComplete := false
+	activityCount := int32(5)
+	activityCounter := int32(0)
+
+	dtHandler := func(execution *types.WorkflowExecution, wt *types.WorkflowType,
+		previousStartedEventID, startedEventID int64, history *types.History) ([]byte, []*types.Decision, error) {
+		if activityCounter < activityCount {
+			activityCounter++
+			buf := new(bytes.Buffer)
+			s.Nil(binary.Write(buf, binary.LittleEndian, activityCounter))
+
+			return []byte(strconv.Itoa(int(activityCounter))), []*types.Decision{{
+				DecisionType: types.DecisionTypeScheduleActivityTask.Ptr(),
+				ScheduleActivityTaskDecisionAttributes: &types.ScheduleActivityTaskDecisionAttributes{
+					ActivityID:                    strconv.Itoa(int(activityCounter)),
+					ActivityType:                  &types.ActivityType{Name: activityName},
+					TaskList:                      &types.TaskList{Name: testTaskListName},
+					Input:                         buf.Bytes(),
+					ScheduleToCloseTimeoutSeconds: common.Int32Ptr(1),
+					ScheduleToStartTimeoutSeconds: common.Int32Ptr(1),
+					StartToCloseTimeoutSeconds:    common.Int32Ptr(1),
+					HeartbeatTimeoutSeconds:       common.Int32Ptr(1),
+				},
+			}}, nil
+		}
+
+		s.Logger.Info("Completing Workflow.")
+
+		workflowComplete = true
+		return []byte(strconv.Itoa(int(activityCounter))), []*types.Decision{{
+			DecisionType: types.DecisionTypeCompleteWorkflowExecution.Ptr(),
+			CompleteWorkflowExecutionDecisionAttributes: &types.CompleteWorkflowExecutionDecisionAttributes{
+				Result: []byte("Done."),
+			},
+		}}, nil
+	}
+
+	activityExecutedCount := int32(0)
+	atHandler := func(execution *types.WorkflowExecution, activityType *types.ActivityType,
+		activityID string, input []byte, taskToken []byte) ([]byte, bool, error) {
+		s.Equal(testWorkflowID, execution.WorkflowID)
+		s.Equal(activityName, activityType.Name)
+
+		activityExecutedCount++
+		return []byte("Activity Result."), false, nil
+	}
+
+	poller := &TaskPoller{
+		Engine:          s.engine,
+		Domain:          s.domainName,
+		TaskList:        &types.TaskList{Name: testTaskListName},
+		Identity:        testIdentity,
+		DecisionHandler: dtHandler,
+		ActivityHandler: atHandler,
+		Logger:          s.Logger,
+		T:               s.T(),
+	}
+
+	for i := int32(0); i < activityCount; i++ {
+		_, err = poller.PollAndProcessDecisionTask(false, false)
+		s.NoError(err)
+
+		err = poller.PollAndProcessActivityTask(false)
+		s.NoError(err)
+	}
+
+	s.Logger.Info("Waiting for workflow to complete", tag.WorkflowRunID(we.RunID))
+
+	s.False(workflowComplete)
+	_, err = poller.PollAndProcessDecisionTask(false, false)
```

Comment:
I'm not super familiar with these engine tests. Does this block here waiting for a reponse (for N seconds or whatever) until there's a task?

Reply:
It has a fixed 5 attempts and then says no tasks found

Reply:
host/taskpoller.go